arch/riscv/core/smp.c

```diff
@@ -72,7 +71,7 @@ void arch_sched_ipi(void)
 
 	for (i = 0U; i < num_cpus; i++) {
 		if (i != id) {
-			volatile uint32_t *r = (uint32_t *)get_hart_msip(i);
+			volatile uint32_t *r = get_hart_msip(_kernel.cpus[i].arch.hartid);
```

Comment:
Will arch_sched_ipi get called before z_riscv_secondary_cpu_init?
if that happens, .arch.hartid will not be set for other harts. get_hart_msip()  will then return the wrong value for the msip registers of the other harts.

Reply:
Indeed, that is a problem. I'll push a change to guard against this.